docs/development/CUSTOM_RULES.md

```diff
@@ -0,0 +1,206 @@
+---
+id: custom-rules
+sidebar_label: Custom Rules
+title: Custom Rules
+---
+
+:::important
+You should be familiar with [ESLint's developer guide](https://eslint.org/docs/developer-guide) and [Development > Architecture](./architecture/asts) before writing custom rules.
+:::
+
+As long as you are using `@typescript-eslint/parser` as the `parser` in your ESLint configuration, custom ESLint rules generally work the same way for JavaScript and TypeScript code.
+The main two changes to custom rules writing are:
+
+- [AST Extensions](#ast-extensions): targeting TypeScript-specific syntax in your rule selectors
+- [Typed Rules](#typed-rules): using the TypeScript type checker to inform rule logic
+
+## AST Extensions
+
+`@typescript-eslint/estree` creates AST nodes for TypeScript syntax with names that begin with `TS`, such as `TSInterfaceDeclaration` and `TSTypeAnnotation`.
+These nodes are treated just like any other AST node.
+You can query for them in your rule selectors.
+
+This rule written in JavaScript bans interfaces that start with a lower-case letter:
+
+```js
+export const rule = {
+  create(context) {
+    return {
+      TSInterfaceDeclaration(node) {
+        if (/\p{Lu}/.test(node.id.name[0])) {
```

Comment:
Wait - what is this regex style? Shouldn't it be `/[a-z]/`?

Reply:
https://unicode.org/reports/tr18/#General_Category_Property

...except, I'm finding it to not be working in my dev tools, so I'll just go ahead and switch.

Reply:
TIL - i'd never seen it before!
I never mess with unicode though heh

Reply:
another way to do this which would let it work independent of latin characters is

```ts
if (node.id.name[0].toUpperCase() === node.id.name[0]) {
```